core/store/src/trie/trie_storage.rs

```diff
@@ -330,27 +332,23 @@ impl TrieStorage for TrieMemoryPartialStorage {
             }))
         }
     }
-
-    fn as_partial_storage(&self) -> Option<&TrieMemoryPartialStorage> {
-        Some(self)
-    }
 }
 
 impl TrieMemoryPartialStorage {
     pub fn new(recorded_storage: HashMap<CryptoHash, Arc<[u8]>>) -> Self {
         Self { recorded_storage, visited_nodes: Default::default() }
     }
 
-    pub fn partial_state(&self) -> PartialState {
-        let touched_nodes = self.visited_nodes.read();
-        let mut nodes: Vec<_> =
-            self.recorded_storage
-                .iter()
-                .filter_map(|(node_hash, value)| {
-                    if touched_nodes.contains(node_hash) { Some(value.clone()) } else { None }
-                })
-                .collect();
-
+    pub fn partial_state(self) -> PartialState {
+        let Self { recorded_storage, visited_nodes } = self;
+        let mut nodes: Vec<_> = recorded_storage
+            .into_iter()
+            .filter_map(
+                |(node_hash, value)| {
+                    if visited_nodes.contains(&node_hash) { Some(value) } else { None }
```

Comment:
nit: not sure if better or worse but alternatively you can write:
`visited_nodes.contains(&node_hash).then(value)` 
or something similar, maybe then_some would be better